### PR TITLE
refactor: evict inactive services in the checkpoint

### PR DESCRIPTION
**Pull request by kupermind**
- Evict inactive services in the checkpoint.

### CODE REVIEW

File: contracts/staking/ServiceStakingBase.sol [hunk 1]
```diff
@@ -109,6 +105,12 @@ struct ServiceInfo {
 /// @author Andrey Lebedev - <andrey.lebedev@valory.xyz>
 /// @author Mariapia Moscatiello - <mariapia.moscatiello@valory.xyz>
 abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
+    enum ServiceStakingState {
+        Unstaked,
+        Staked,
+        Evicted
+    }
```
Comment (kupermind) [inline contracts/staking/ServiceStakingBase.sol:112]: Giving a bit more information about service staking statuses

File: contracts/staking/ServiceStakingBase.sol [hunk 2]
```diff
@@ -435,20 +440,81 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+    function _evict(uint256[] memory evictServiceIds) internal {
+        uint256 totalNumServices = evictServiceIds.length;
+        uint256 numEvictServices;
+
+        // Get the number of evicted services
+        for (uint256 i = 0; i < totalNumServices; ++i) {
+            if (evictServiceIds[i] > 0) {
+                numEvictServices++;
+            }
+        }
```
Comment (kupermind) [inline contracts/staking/ServiceStakingBase.sol:454]: The input array is always of a size of all the staked services, and thus for the event exact number of evicted services must be calculated.

File: contracts/staking/ServiceStakingBase.sol [hunk 3]
```diff
@@ -435,20 +440,81 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+            for (uint256 i = 0; i < totalNumServices; ++i) {
+                if (evictServiceIds[i] > 0) {
+                    serviceId = evictServiceIds[i];
+                    serviceIds[sCounter] = serviceId;
+
+                    ServiceInfo storage sInfo = mapServiceInfo[serviceId];
+                    owners[sCounter] = sInfo.owner;
+                    multisigs[sCounter] = sInfo.multisig;
+                    serviceInactivity[sCounter] = sInfo.inactivity;
+                    serviceIndexes[sCounter] = i;
```
Comment (kupermind) [inline contracts/staking/ServiceStakingBase.sol:476]: Get all the data of each evicted service.

File: contracts/staking/ServiceStakingBase.sol [hunk 4]
```diff
@@ -435,20 +440,81 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+                    sCounter++;
+                }
+            }
+
+            // Evict services from the global set of staked services
+            uint256 idx;
+            for (uint256 i = numEvictServices - 1; i > 0; --i) {
```
Comment (kupermind) [inline contracts/staking/ServiceStakingBase.sol:483]: When evicting, we go backwards, as if we go forward all the indexes will be changed with every deletion.

File: contracts/staking/ServiceStakingBase.sol [hunk 5]
```diff
@@ -457,9 +523,16 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
             uint256[] memory serviceIds, uint256[][] memory serviceNonces,
             uint256[] memory serviceInactivity) = _calculateStakingRewards();
 
+        // Get arrays for eligible service Ids and rewards of exact size
+        uint256[] memory finalEligibleServiceIds;
+        uint256[] memory finalEligibleServiceRewards;
+        evictServiceIds = new uint256[](serviceIds.length);
+        uint256 curServiceId;
```
Comment (kupermind) [inline contracts/staking/ServiceStakingBase.sol:530]: For the event, we need a precise number of eligible services.

Also, evict services are allocated to the size of the all services in the set of staked services.

File: contracts/staking/ServiceStakingBase.sol [hunk 6]
```diff
@@ -505,30 +584,43 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
                 if (serviceInactivity[i] > 0) {
-                    mapServiceInfo[curServiceId].inactivity += serviceInactivity[i];
+                    // Get the overall continuous service inactivity
+                    uint256 inactivity = mapServiceInfo[curServiceId].inactivity + serviceInactivity[i];
+                    mapServiceInfo[curServiceId].inactivity = inactivity;
+                    // Check for the maximum allowed inactivity time
+                    if (inactivity > maxAllowedInactivity) {
+                        // Evict a service if it has been inactive for more than a maximum allowed inactivity time
+                        evictServiceIds[i] = curServiceId;
+                    }
```
Comment (kupermind) [inline contracts/staking/ServiceStakingBase.sol:603]: If inactivity is bigger than the allowed one, the service assigned to evicted services list.

File: contracts/staking/ServiceStakingBase.sol [hunk 7]
```diff
@@ -540,26 +635,36 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
             revert OwnerOnly(msg.sender, sInfo.owner);
         }
 
-        // Check that the service has staked long enough, or if there are no rewards left
+        // Get the staking start time
+        // Note that if the service info exists, the service is staked or evicted, and thus start time is always valid
         uint256 tsStart = sInfo.tsStart;
+
+        // Check that the service has staked long enough, or if there are no rewards left
```
Comment (mariapiamo) [inline contracts/staking/ServiceStakingBase.sol:645]: we still need to discuss about the possible issue on code updates here

**Review by mariapiamo — COMMENTED** [review]
Can we have 
1. a test showing that if maxNumStaker is evicted, after the checkpoint call, a new service can stake? 
2. Even if not specifically related to this pr, a test showing the revert of UnauthorizedMultisig(service.multisig)

File: contracts/staking/ServiceStakingBase.sol (hunk 7 above)
Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:645]: Not sure I understand from description

Comment (mariapiamo) [reply contracts/staking/ServiceStakingBase.sol:645]: https://github.com/valory-xyz/autonolas-registries/pull/139#discussion_r1394762113

**Comment by kupermind** [comment]
@mariapiamo I've added requested tests

File: test/ServiceStaking.js [hunk 8]
```diff
@@ -1167,6 +1185,61 @@ describe("ServiceStaking", function () {
...
+            await serviceRegistry.activateRegistration(deployer.address, serviceId + 2, {value: regDeposit});
+            await serviceRegistry.registerAgents(operator.address, serviceId + 2, [agentInstances[2].address], agentIds, {value: regBond});
+            await serviceRegistry.deploy(deployer.address, serviceId + 2, gnosisSafeMultisig.address, payload);
+
+            // Stake 3 initial services
+            for (let i = 0; i < 3; i++) {
+                // Approve services
+                await serviceRegistry.approve(serviceStaking.address, serviceId + i);
+
+                // Stake the first service
```
Comment (mariapiamo) [inline test/ServiceStaking.js:1206]: // stake the service

**Review by mariapiamo — COMMENTED** [review]
@kupermind thanks! Just a minor suggestion for a comment

File: contracts/staking/ServiceStakingBase.sol [hunk 9]
```diff
@@ -131,11 +133,13 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
         bytes32 configHash;
     }
 
-    event ServiceStaked(uint256 indexed serviceId, address indexed owner, address indexed multisig, uint256[] nonces);
-    event Checkpoint(uint256 availableRewards, uint256 numServices);
-    event ServiceUnstaked(uint256 indexed serviceId, address indexed owner, address indexed multisig, uint256[] nonces,
-        uint256 reward, uint256 tsStart);
-    event ServiceEvicted(uint256 indexed serviceId, address indexed owner, address indexed multisig, uint256 inactivity);
+    event ServiceStaked(uint256 indexed epoch, uint256 indexed serviceId, address indexed multisig, address owner,
```
Comment (DavidMinarsch) [inline contracts/staking/ServiceStakingBase.sol:136]: why is owner not indexed?

File: contracts/staking/ServiceStakingBase.sol [hunk 10]
```diff
@@ -131,11 +133,13 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+        uint256[] nonces);
+    event Checkpoint(uint256 indexed epoch, uint256 availableRewards, uint256[] serviceIds, uint256[] rewards);
+    event ServiceUnstaked(uint256 indexed epoch, uint256 indexed serviceId, address indexed multisig, address owner,
```
Comment (DavidMinarsch) [inline contracts/staking/ServiceStakingBase.sol:139]: see above

why tsStart? What's the point of emitting this here?

File: contracts/staking/ServiceStakingBase.sol [hunk 11]
```diff
@@ -131,11 +133,13 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+        uint256[] nonces, uint256 reward, uint256 tsStart);
+    event ServicesEvicted(uint256 indexed epoch, uint256[] serviceIds, address[] owners, address[] multisigs,
```
Comment (DavidMinarsch) [inline contracts/staking/ServiceStakingBase.sol:141]: order of multisigs and owners need to be reverted to be symmetric with the other events

File: contracts/staking/ServiceStakingBase.sol [hunk 12]
```diff
@@ -435,20 +442,80 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+
     /// @dev Checkpoint to allocate rewards up until a current time.
-    /// @return All staking service Ids.
-    /// @return All staking updated nonces.
-    /// @return Number of reward-eligible staking services during current checkpoint period.
-    /// @return Eligible service Ids.
-    /// @return Eligible service rewards.
-    /// @return success True, if the checkpoint was successful.
+    /// @return All staking service Ids (including evicted ones during within a current epoch).
+    /// @return All staking updated nonces (including evicted ones during within a current epoch).
```
Comment (DavidMinarsch) [inline contracts/staking/ServiceStakingBase.sol:509]: during within?

File: contracts/staking/ServiceStakingBase.sol [hunk 13]
```diff
@@ -435,20 +442,80 @@ abstract contract ServiceStakingBase is ERC721TokenReceiver, IErrorsRegistries {
...
+    /// @return Set of eligible service Ids.
```
Comment (DavidMinarsch) [inline contracts/staking/ServiceStakingBase.sol:510]: reward-eligible

Comment (DavidMinarsch) [reply contracts/staking/ServiceStakingBase.sol:510]: why do we no longer return the nonces?

File: contracts/staking/ServiceStakingBase.sol (hunk 9 above)
Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:136]: It's max of 3 indexed fields. I've changed that now, not indexing the epoch.

File: contracts/staking/ServiceStakingBase.sol (hunk 13 above)
Comment (kupermind) [reply contracts/staking/ServiceStakingBase.sol:510]: We return nonces, we don't return the success flag. No need to return that, just can check for the `serviceIds` size.

**Review by mariapiamo — APPROVED** [review]

**Pull request merged**